src/daft-sql/src/modules/partitioning.rs

```diff
@@ -1,11 +1,96 @@
+use daft_dsl::functions::partitioning::{self, PartitioningExpr};
+
 use super::SQLModule;
-use crate::functions::SQLFunctions;
+use crate::{
+    ensure,
+    functions::{SQLFunction, SQLFunctions},
+};
 
 pub struct SQLModulePartitioning;
 
 impl SQLModule for SQLModulePartitioning {
-    fn register(_parent: &mut SQLFunctions) {
-        // use FunctionExpr::Partitioning as f;
-        // TODO
+    fn register(parent: &mut SQLFunctions) {
+        parent.add_fn("partitioning_years", PartitioningExpr::Years);
+        parent.add_fn("partitioning_months", PartitioningExpr::Months);
+        parent.add_fn("partitioning_days", PartitioningExpr::Days);
+        parent.add_fn("partitioning_hours", PartitioningExpr::Hours);
+        parent.add_fn(
+            "partitioning_iceberg_bucket",
+            PartitioningExpr::IcebergBucket(0),
+        );
+        parent.add_fn(
+            "partitioning_iceberg_truncate",
+            PartitioningExpr::IcebergTruncate(0),
+        );
+    }
+}
+
+impl SQLFunction for PartitioningExpr {
+    fn to_expr(
+        &self,
+        args: &[sqlparser::ast::FunctionArg],
+        planner: &crate::planner::SQLPlanner,
+    ) -> crate::error::SQLPlannerResult<daft_dsl::ExprRef> {
+        fn partitioning_helper<F: FnOnce(daft_dsl::ExprRef) -> daft_dsl::ExprRef>(
+            args: &[sqlparser::ast::FunctionArg],
+            planner: &crate::planner::SQLPlanner,
+            method_name: &str,
+            f: F,
+        ) -> crate::error::SQLPlannerResult<daft_dsl::ExprRef> {
+            ensure!(args.len() == 1, "{} takes exactly 1 argument", method_name);
+            let args = planner.plan_function_arg(&args[0])?;
+            Ok(f(args))
+        }
+        match self {
+            PartitioningExpr::Years => {
+                partitioning_helper(args, planner, "years", partitioning::years)
+            }
+            PartitioningExpr::Months => {
+                partitioning_helper(args, planner, "months", partitioning::months)
+            }
+            PartitioningExpr::Days => {
+                partitioning_helper(args, planner, "days", partitioning::days)
+            }
+            PartitioningExpr::Hours => {
+                partitioning_helper(args, planner, "hours", partitioning::hours)
+            }
+            PartitioningExpr::IcebergBucket(_) => {
+                let input = planner.plan_function_arg(&args[0])?;
```

Comment:
Could `args` potentially be an empty slice here? And therefore, `&args[0]` would fail.

Reply:
(Unless we're checking before hand that `args` is not empty for the `IcebergBucket` case).

Reply:
good catch, we should definitely check the num args first!